### PR TITLE
Fix support for processors

### PR DESCRIPTION
**Pull request by gucong3000**
> Which issue, if any, is this issue related to?

Closes: #3252
Closes: https://github.com/styled-components/stylelint-processor-styled-components/issues/180

> Is there anything in the PR that needs further explanation?

No, it's self explanatory.

### CODE REVIEW

File: docs/user-guide/processors.md [hunk 1]
```diff
@@ -13,6 +13,11 @@ Processors are community packages that enable stylelint to lint the CSS within n
 
 ## PostCSS syntax
```
Comment (jeddy3) [inline docs/user-guide/processors.md]: Let's just remove this whole section for now. This is the `processors.md` file. Once we resolve the relationship between processors and syntax we can revisit this.

File: docs/user-guide/css-processors.md [hunk 2]
```diff
@@ -17,10 +17,10 @@ By default, the linter can *parse* any the following non-standard syntaxes by us
 -   SCSS (using [`postcss-scss`](https://github.com/postcss/postcss-scss))
 -   Less (using [`postcss-less`](https://github.com/shellscape/postcss-less))
 -   SugarSS (using [`sugarss`](https://github.com/postcss/sugarss))
--   HTML,
```
Comment (jeddy3) [inline docs/user-guide/css-processors.md]: Let's remove "Javascript or typescript" from this list for now as it's an undocumented feature. And let's remove "Markdown" and "HTML-like" as we're still resolving the processor vs syntax issue.

**Comment by jeddy3** [comment]
@gucong3000 Many thanks for starting on this!

Also, thanks for [splitting the parsers into their own packages](https://github.com/stylelint/stylelint/issues/3248#issuecomment-381385995). I'm confident this will help down the line.

Sorry for the delay reviewing this PR. The changes are far reaching and there is a lot to understand. 

I've tested this branch with all the [processors listed in the docs](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/processors.md):

- [x] [@mapbox/stylelint-processor-arbitrary-tags](https://github.com/mapbox/stylelint-processor-arbitrary-tags)
- [x] [stylelint-processor-glamorous](https://github.com/zabute/stylelint-processor-glamorous)
- [ ] [@mapbox/stylelint-processor-markdown](https://github.com/mapbox/stylelint-processor-markdown)
- [x] [stylelint-processor-styled-components](https://github.com/styled-components/stylelint-processor-styled-components)

The good news is most, including `stylelint-processor-styled-components`, are compatible with this branch. _However, `@mapbox/stylelint-processor-markdown` is not._ 

Given the following config:

```json
{
  "processors": ["@mapbox/stylelint-processor-markdown"],
  "extends": [
    "stylelint-config-standard"
  ]
}
```

And the following markdown:

````
Uh oh, I'm about to use SCSS.

```scss
// Parser-breaking comment
$foo: bar;
.foo {}
```
````

The following error is thrown:

```console
jeddy3:stylelint-test jeddy3$ ./node_modules/.bin/stylelint "sources/md/scss.*" --config configs/markdown.json
TypeError: Cannot read property 'line' of undefined
    at result.warnings.reduce (/Users/jeddy3/Projects/stylelint-test/node_modules/@mapbox/stylelint-processor-markdown/index.js:54:39)
    at Array.reduce (<anonymous>)
    at transformResult (/Users/jeddy3/Projects/stylelint-test/node_modules/@mapbox/stylelint-processor-markdown/index.js:51:39)
    at config.resultProcessors.forEach.resultProcessor (/Users/jeddy3/Projects/stylelint/lib/createStylelintResult.js:174:26)
    at Array.forEach (<anonymous>)
    at stylelint.getConfigForFile.then (/Users/jeddy3/Projects/stylelint/lib/createStylelintResult.js:171:31)
    at <anonymous>
```

It is the `scss` (or `less`or `sass`) on the code fence that breaks it. Using `css` is compatible.

**This _regression_ in this branch will need to be fixed before we can merge this PR.**

Then, after a patch is released, I believe stylelint will once again be compatible with the existing processors. We can then, as we prepare for `stylelint@10.0`, carefully deliberate on:

1. the role of processors and syntax
2. what features (e.g. markdown, `<style>`, styled components) stylelint should support _by default_.

**Comment by gucong3000** [comment]
> However, @mapbox/stylelint-processor-markdown is not.

Still fail when I remove `syntax` from postcssOptions:
https://github.com/stylelint/stylelint/blob/issue-3248/lib/getPostcssResult.js#L92

**Comment by jeddy3** [comment]
@gucong3000 Thanks for making the changes and for looking into the markdown processor regression. Sorry for the slow response. I'm just catching up on 2 weeks of stylelint issues.

Ok, I think we should merge this PR and get the release out. It's more important to fix the styled-components processor regression because users might be relying on the interpolation tagging feature of the processor. Anyone using the markdown-processor (with non-standard syntax code fences) can simply remove processor and I believe they'll get the same functionality. Sound good?

**Comment by stramel** [comment]
@jeddy3 any update on this?

**Comment by jeddy3** [comment]
It would be great to see this merged.  @stylelint/core A second pair of eyes on it would be great, though.

If no one has time, I'll try to make some to test again this evening and get this merged.

**Comment by hudochenkov** [comment]
I'm not sure what this PR does in general. Does it introduce breaking change?

Code looks good.

File: lib/getPostcssResult.js [hunk 3]
```diff
@@ -1,28 +1,29 @@
...
-  scss,
-  sss: sugarss,
-  sugarss
+  }
 };
+[
+  "less",
+  "sass",
+  "scss",
+  "markdown",
```
Comment (ntwb) [inline lib/getPostcssResult.js]: `markdown` was being removed wasn't it?

File: package.json [hunk 4]
```diff
@@ -62,15 +62,17 @@
     "normalize-selector": "^0.2.0",
     "pify": "^3.0.0",
     "postcss": "^6.0.16",
-    "postcss-html": "^0.18.0",
+    "postcss-html": "^0.23.6",
     "postcss-less": "^1.1.5",
+    "postcss-markdown": "^0.23.6",
```
Comment (ntwb) [inline package.json]: Likewise, wasn't `markdown` being removed for now, in this PR and to be added back in a PR of its own?

File: package.json [hunk 5]
```diff
@@ -113,7 +115,7 @@
     "precommit": "lint-staged",
     "benchmark-rule": "node scripts/benchmark-rule.js",
     "dry-release": "npmpub  --dry --verbose",
-    "flow": "flow",
+    "flow": "echo flow",
```
Comment (ntwb) [inline package.json]: Why `echo` the `flow` word instead of running `flow`?

File: package.json (hunk 4 above)
Comment (gucong3000) [reply package.json]: Previously Markdown's processing logic was written in `postcss-html`, and now it is split into multiple packages: `postcss-syntax`, `postcss-html` and `postcss-markdown`

File: lib/getPostcssResult.js (hunk 3 above)
Comment (gucong3000) [reply lib/getPostcssResult.js]: `postcss-markdown` split from the original `postcss-html`

**Comment by jeddy3** [comment]
> -   Fixed: regression with processors (e.g. styled-components) ([#3261](https://github.com/stylelint/stylelint/pull/3261)).

**Pull request merged**